### ISSUE TITLE
RTP issue--"Incorrect timestamp" in packet trace (voice quality problems)

### ISSUE THREAD

**Issue by richard-tkg** [signalwire/freeswitch#1651]
Summary:  large numbers of RTP packets show "Incorrect timestamp" in Wireshark

Details:
Freeswitch version 1.10.7 -release-19-883d2cb662 64bit

Call is placed from a Zoiper client to a cellular telephone.
Zoiper => Freeswitch => Bandwidth.com trunk/AT&T cellular telephone

The RTP streams are good in all directions except from Freeswitch to the Zoiper client (Stream 0).  Stream 1 from Zoiper to Freeswitch shows no issues.  A screenshot of the Wireshark RTP Stream Analysis is attached.

The issue occurs on all calls we have analyzed.  

I can provide more details as needed.  Any suggestions would be much appreciated as this has been a rather vexing issue for us, and our clients will be very happy when the issue is resolved.

![image](https://user-images.githubusercontent.com/18629568/167748166-7c74a140-03ea-4d41-94ed-a613d2eef085.png)

**Comment by mjerris** [comment]
did you verify those messages are actually correct?  for years i’ve seen that analyzer show stuff like that that was incorrect

**Comment by mjerris** [comment]
need a full log with debug and sip trace and a full pcap of both legs of the call with media to look at this issue at all

**Comment by richard-tkg** [comment]
Thanks much. I will check to see if the packets are actually broken and will send the additional debug information.

**Comment by jgarland79** [comment]
<img width="1803" alt="Screen Shot 2022-05-10 at 10 24 36 PM" src="https://user-images.githubusercontent.com/1667027/168132936-37470fef-9b6f-4dab-9cbf-7fbe0e076ebf.png">

I've been seeing these errors occasionally in voipmonitor. Do we need to set up ntp on our servers?

@Null-Is-Null

**Comment by shaunjstokes** [comment]
This looks similar to a problem we recently came across, check if you're using `bridge_generate_comfort_noise`, the FS generated comfort RTP timestamps won't match the original RTP timestamps.

**Comment by shaunjstokes** [comment]
This appears to be the same issue we are experiencing.
![image](https://user-images.githubusercontent.com/14235585/178446063-09cd284e-afa6-4669-a7aa-1d34dc9ec3ce.png)

It only occurs on missed packets when using record_session or uuid_record. When there is a missed packet on Leg A, FS generates the missing packet on Leg B using a different timestamp.

The issue is resolved in version 1.10.1 and is present from version 1.10.4, I haven't been able to test 1.10.2 or 1.10.3 due to build problems.

**Comment by shaunjstokes** [comment]
The issue is caused by this commit: https://github.com/signalwire/freeswitch/commit/47c5c8f3e8957c37fad5feeeb791375d05992b93

To resolve the issue, we simply need to re-apply the following line in 'src/switch_rtp.c'.
if (!ts) ts = rtp_session->last_write_ts + rtp_session->samples_per_interval;

The above statement is triggered for FS generated RTP packets on an active session, and will result in FS using next RTP timestamp instead of generating a new timestamp, surely this is correct? Why generate a new timestamp when we know the next timestamp?

**Comment by shaunjstokes** [comment]
The change was authored by @anthmFS and committed by @andywolk , can you please explain why this line was removed?

**Comment by greenbea** [comment]
I've seen the incorrect timestamps in my captures way before version 1.10.

**Comment by shaunjstokes** [comment]
@richard-tkg we were able to reproduce the issue consistently once we understood what triggered it. That's correct, the problem does not exist in 1.10.1, but going back that far isn't an option for us as it introduces other issues.

I've attached the patch that we're using. To use the patch, just unzip it into your FS source directory, then use `git apply switch_rtp_use_last_write_ts.diff` before the `make` command when building from source.

[switch_rtp_use_last_write_ts.zip](https://github.com/signalwire/freeswitch/files/9118718/switch_rtp_use_last_write_ts.zip)

I wanted to get some feedback from the FS Devs before I go ahead and create a pull request.

**Comment by richard-tkg** [comment]
@shaunjstokes On our existing servers, Freeswitch was installed from Debian packages.  I have also built FS from source and applied your patch.  I'd like to keep the existing package-based installation as-is and modify only as few files as necessary to apply the newly built version of Freeswitch.  Have you done this before?  Any idea if it's possible and if so, is it advisable?

**Comment by shaunjstokes** [comment]
@richard-tkg it's compiled, it would be best to install the source code version to a separate directory, apply your configuration, then update the FS service to use the source code version. Initially we used the packaged version of FreeSWITCH, but we found the source code version to be much more versatile, we've never looked back.

**Comment by emaktech** [comment]
So it seems that this line was removed after this discussion: https://github.com/signalwire/freeswitch/issues/146

Can anybody confirm adding this line back fixes the issue described here? Because we have this too and it's a huge issue whenever there's dropped packets or jitter on the A leg!

**Comment by shaunjstokes** [comment]
Well spotted @emaktech, so the final patch needs to determine "when files are being played back" as mentioned by @htrendev in #146. I'm not yet sure how we can achieve that, perhaps there's someone here that can help?

**Comment by emaktech** [comment]
Indeed, it seems that at present we need to pick between timestamp issues on phrase playback or timestamp issues with lost/dropped packets :(

FYI I found the original commit by @mjerris [here](https://github.com/signalwire/freeswitch/commit/9e3da931e519c2dbba2f86762756b1f26f8318d8). Obviously this issue had been reported and patched in the past, but unfortunately caused https://github.com/signalwire/freeswitch/issues/146

Somebody please send help!

**Comment by emaktech** [comment]
So we have made some progress on this. During FreeSWITCH office hours today, Brian West pointed me toward some undocumented dialplan variables that can be set. Or rather, they are documented, but only in the source code here:

https://github.com/signalwire/freeswitch/blob/9df85738e10d80a36c4f84807fbc30c8e831ee11/src/include/switch_types.h#L969

There are two variables that you can try setting and see if it helps:

You can try `RTP_BUG_SEND_LINEAR_TIMESTAMPS` and `RTP_BUG_GEN_ONE_GEN_ALL` (and maybe others, read through the list starting at line 856 and see what you find).

Basically do a set (or export) `rtp_manual_rtp_bugs=RTP_BUG_GEN_ONE_GEN_ALL` or any of the others above and see if it helps!

**Comment by shaunjstokes** [comment]
@emaktech some of those variables might fix the issue, but they also go further than what's required, we did testing with various RTP bug variables before coming here.

The problem is very specific, when there are missing RTP packets on Leg A, we still want the RTP packets generated by FreeSWITCH on Leg B to use the next timestamp which is logical. Otherwise, FreeSWITCH should pass on the RTP timestamp from Leg A to Leg B, so the patch is still the best option for us.

**Comment by Kobyhud** [comment]
We can confirm this issue. Building from source and making the change recommended by shaunjstokes does seem to solve the issue that we experience. In our case it is manifested only on uuid_recorded outbound channels. Certain endpoints really cannot seem to deal with the timestamp re-sequences very well at all. In our case the problematic endpoint is Counterpath's Bria. 
Based on our testing shaunjstokes patch doesn't have any effect on phrases. We see RTP timestamp issues both before and after this patch.

**Comment by olofssonanton** [comment]
I can confirm that recorded sessions still suffer from this issue on v1.10.11, and reverting https://github.com/signalwire/freeswitch/commit/47c5c8f3e8957c37fad5feeeb791375d05992b93 as suggested by @shaunjstokes resolves the issue.

**Comment by olofssonanton** [comment]
Although the real question is why are there packages being generated at all?

Inserting packages into the audio stream as soon as there are seemingly missing packets will negate the effect of any jitter buffer on the receiving end, while simultaneously distorting the audio.

I assume it is done for the sake of the recording, but this should be possible without affecting the packets written to the receiver. Perhaps even without introducing any distortion into the recording either, by re-ordering packets in a jitter buffer before writing to the file?

**Comment by zusrut** [comment]
when do `uuid_record` with `RECORD_READ_ONLY=true` for per leg recording - on any jitter in recording freeswitch inputs silence and then when packet appears it append it after this extra silence. so leg recordings have different duration and after some duration starts overlapping.
here example with this stats:
```
in_jitter_packet_count:0
in_dtmf_packet_count:0
in_cng_packet_count:0
in_flush_packet_count:27
in_largest_jb_size:0
in_jitter_min_variance:110.50
in_jitter_max_variance:474.53
in_jitter_loss_rate:0.00
in_jitter_burst_rate:1.00
in_mean_interval:20.50
in_flaw_total:1123
```


and got such picture:

![image](https://github.com/user-attachments/assets/90d9aeb7-4b62-4e76-8ea0-11d7fe886406)

for better jitter the picture is better but bad anyway .
increase of param `auto-jitterbuffer-msec` helps but with own cons. issue still can appear on worse quality calls

**Comment by hashcode86** [comment]
After revert, i faced problem as mentioned in [#146](https://github.com/signalwire/freeswitch/issues/146)

RTP packet 1,   seq=1, time=0
RTP packet 2,   seq=2, time=160
--- 120ms of no RTP ---
RTP packet 3,   seq=3, time=320 (timestamp is seemingly wrong)
RTP packet 4,   seq=4, time=480

My workaround solution:
- remove: if (!ts) ts = rtp_session->last_write_ts + rtp_session->samples_per_interval;
- enable rtp-rewrite-timestamps: <param name="rtp-rewrite-timestamps" value="true"/>
